Update arch documentation with new diagram, descriptions (#1329)
nasa/cumulus: docs/architecture.md

hide_title: false
---

## Architecture

Below, find a diagram with the components that comprise an instance of Cumulus.

![Architecture diagram of a Cumulus deployment](assets/cumulus-arch-diagram.png)

This diagram details all of the major architectural  components of a Cumulus deployment.

While the diagram can feel complex, it can easily be digested in several major components:

### Data Distribution

End Users can access data via Cumulus's `distribution` submodule, which includes ASF's [thin egress application](https://github.com/asfadmin/thin-egress-app), this provides authenticated data egress, temporary S3 links and other statistics features.

### Data ingest

#### Workflows

The core of the ingest and processing capabilities in Cumulus is built into the deployed AWS [Step Function](https://aws.amazon.com/step-functions/) workflows.    Cumulus rules trigger workflows via either Cloud Watch rules, Kinesis streams, SNS topic, or SQS queue.   The workflows then run with a configured [Cumulus message](./workflows/cumulus-task-message-flow), utilizing built-in processes to report status of granules, PDRs, executions, etc to the [Data Persistence](#data-persistence) components.

> Workflows can optionally report granule status to [CMR](<https://earthdata.nasa.gov/eosdis/science-system-description/eosdis-components/cmr>), and if configured and using the [Cumulus message adapter (CMA)](<https://github.com/nasa/cumulus-message-adapter>), workflow steps can report metrics information to a shared SNS topic, for export to an external ELK stack to make use of.

Reviewer: ```suggestion
Workflows can optionally report granule metadata to [CMR](<https://earthdata.nasa.gov/eosdis/science-system-description/eosdis-components/cmr>), and if configured to use the [Cumulus message adapter (CMA)](<https://github.com/nasa/cumulus-message-adapter>), workflow steps can report metrics information to a shared SNS topic, which could be subscribed to for near real time granule, execution, and PDR status. This could be used for metrics reporting using an external ELK stack, for example.
```

Reviewer: So you have to use the CMA to have it post to the SNS topic?

Author: @laurenfrederick my understanding is the cloudwatch logs won't have the metadata needed for stats reporting if the CMA is not involved.   @flamingbear ?

Reviewer 2: lauren did those tickets.  But my understanding is cloudwatch logs will need the CMA enabled in order to get the enhanced logs from the lambda tasks.  But I don't think that's the case for the SNS "logs" (messages send to sns), I think those are standalone.  I haven't looked at them yet though.

Author: Hmm, I see, I think I have a misunderstanding of what that portion of the diagram implies.     Sounds like we should pull out the portion about the CMA?

Reviewer: Yeah, the CMA is not required for that as far as i know. It's based on the output of the workflow.

#### Data persistence

Cumulus entity state data is stored in a set of [DynamoDB](https://aws.amazon.com/dynamodb/) database tables, and is exported to an ElasticSearch instance for non-authoritative querying/state data for the API and other applications that require more complex queries.

#### Data discovery

> Data discovery is assumed to be provided via external service.  For NASA use, this is assumed to be [CMR](<https://earthdata.nasa.gov/eosdis/science-system-description/eosdis-components/cmr>).

Reviewer: Data discovery is where we discover granules and PDRs, so the provider types. S3, HTTP/HTTPs, FTP.

Author: Good catch, this is an amalgam of edits gone wrong.  Will fix.

Author: 5effb02f7

### Maintenance

System maintenance personnel have access to manage ingest and various portions of Cumulus via an [AWS API gateway](<https://aws.amazon.com/api-gateway/>), as well as the operator [dashboard](https://github.com/nasa/cumulus-dashboard).

## Deployment Structure

Cumulus is deployed via [Terraform](https://www.terraform.io/) and is organized internally into two seperate top-level modules, as well as several external modules.

### Cumulus

The [Cumulus module](https://github.com/nasa/cumulus/tree/master/tf-modules/cumulus), which contains multiple internal submodules, deploys all of the Cumulus components that are not part of the `Data Persistence` portion of this diagram.

### Data persistence

The [data persistence](https://github.com/nasa/cumulus/tree/master/tf-modules/data-persistence) module provides the `Data Persistence` portion of the diagram.

### Other modules

> Other modules provided on the [release](https://github.com/nasa/cumulus/releases) page are provided for use in users configuring their own deployment and contain extracted subcomponents of the [cumulus](#cumulus) module.  For more on these components see the [components documentation](./deployments/components).

Reviewer: ```suggestion
Other modules are provided as artifacts on the [release](https://github.com/nasa/cumulus/releases) page for use in users configuring their own deployment and contain extracted subcomponents of the [cumulus](#cumulus) module.  For more on these components see the [components documentation](./deployments/components).
```

Reviewer: The components link does not work

Author: be02d9b49

Author: ...and 8aab518fd

> For more on the specific structure, examples of use and how to deploy and more, please see the [deployment](./deployment/README) docs as well as the [cumulus-template-deploy](https://github.com/nasa/cumulus-template-deploy) repo.

Reviewer: Deployment docs link does not work

Author: b7cc74404

While the diagram can feel complex, it can easily be digested in several major components:

### Data Distribution

End Users can access data via Cumulus's `distribution` submodule, which includes ASF's [thin egress application](https://github.com/asfadmin/thin-egress-app), this provides authenticated data egress, temporary S3 links and other statistics features.

#### Data search

End user exporsure of Cumulus's holdings is expected to be provided by an external service.

For NASA use, this is assumed to be [CMR](<https://earthdata.nasa.gov/eosdis/science-system-description/eosdis-components/cmr>) in this diagram.

### Data ingest

#### Workflows

The core of the ingest and processing capabilities in Cumulus is built into the deployed AWS [Step Function](https://aws.amazon.com/step-functions/) workflows.    Cumulus rules trigger workflows via either Cloud Watch rules, Kinesis streams, SNS topic, or SQS queue.   The workflows then run with a configured [Cumulus message](./workflows/cumulus-task-message-flow), utilizing built-in processes to report status of granules, PDRs, executions, etc to the [Data Persistence](#data-persistence) components.

> Workflows can optionally report granule metadata to [CMR](<https://earthdata.nasa.gov/eosdis/science-system-description/eosdis-components/cmr>), and if configured to use the [Cumulus message adapter (CMA)](<https://github.com/nasa/cumulus-message-adapter>), workflow steps can report metrics information to a shared SNS topic, which could be subscribed to for near real time granule, execution, and PDR status. This could be used for metrics reporting using an external ELK stack, for example.

Author: ```suggestion
Workflows can optionally report granule metadata to [CMR](<https://earthdata.nasa.gov/eosdis/science-system-description/eosdis-components/cmr>), and workflow steps can report metrics information to a shared SNS topic, which could be subscribed to for near real time granule, execution, and PDR status. This could be used for metrics reporting using an external ELK stack, for example.
```
